File: LeetCode/2040_H_Kth_Smallest_Product_Of_Two_Sorted_Arrays.c
// #include <stdio.h>
// #include <stdlib.h>
// #include <limits.h>

// void add(long long* arr, int k, long long val) {
//     int indexToInsert = 0;
//     for (int i = 0; i < k; i++) {if (arr[i] < val){indexToInsert++;}}
//     for (int i = k - 1; i > indexToInsert; i--){arr[i] = arr[i - 1];}
//     arr[indexToInsert] = val;
// }

// long long kthSmallestProduct(int* nums1, int nums1Size, int* nums2, int nums2Size, int k) {
//     long long res[k];
//     for (int i = 0; i < k; i++){res[i] = LLONG_MAX;}
//     for (int i = 0; i < nums1Size; i++) {
//         for (int j = 0; j < nums2Size; j++) {
//             long long product = (long long)nums1[i] * (long long)nums2[j];
//             if (product < res[k - 1]){add(res, k, product);}
//         }
//     }
//     return res[k - 1];
// }


#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

long long countLessOrEqual(int* nums1, int nums1Size, int* nums2, int nums2Size, long long target) {
    long long count = 0;
    for (int i = 0; i < nums1Size; i++) {
        if (nums1[i] < 0) { // For negative numbers
            int left = 0, right = nums2Size;
            while (left < right) {
                int mid = left + (right - left) / 2;
                if ((long long)nums1[i] * nums2[mid] <= target) {
                    right = mid;
                } else {
                    left = mid + 1;
                }
            }
            count += nums2Size - right;
        } else if (nums1[i] > 0) { // For positive numbers
            int left = 0, right = nums2Size;
            while (left < right) {
                int mid = left + (right - left) / 2;
                if ((long long)nums1[i] * nums2[mid] <= target) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            count += left;
        } else { // For zero
            if (target >= 0) {
                count += nums2Size;
            }
        }
    }
    return count;
}

long long kthSmallestProduct(int* nums1, int nums1Size, int* nums2, int nums2Size, int k) {
    // Determine the search space for products
    long long left = (long long)nums1[0] * nums2[0];
    long long right = (long long)nums1[nums1Size - 1] * nums2[nums2Size - 1];

    for (int i = 0; i < nums1Size; i++) {
        if ((long long)nums1[i] * nums2[0] < left) left = (long long)nums1[i] * nums2[0];
        if ((long long)nums1[i] * nums2[nums2Size - 1] > right) right = (long long)nums1[i] * nums2[nums2Size - 1];
    }

    long long result = 0;

    // Perform binary search to find the k-th smallest product
    while (left <= right) {
        long long mid = left + (right - left) / 2;
        if (countLessOrEqual(nums1, nums1Size, nums2, nums2Size, mid) >= k) {
            result = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    return result;
}

File: LeetCode/2040_H_Kth_Smallest_Product_Of_Two_Sorted_Arrays.py
# from typing import List
# class Solution:
#     def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
#         remIdxs={i:[j for j in range(len(nums2))] for i in range(len(nums1))}
#         lastK=-1
#         while k:
#             addedIdx,minIdx,minVal=-1,-1,float('inf')
#             for i in range(len(nums1)):
#                 if len(remIdxs[i])>=1:
#                     if nums1[1]>=0:
#                         prod=nums1[i]*nums2[remIdxs[i][0]]
#                         if prod<minVal:
#                             minVal=prod
#                             minIdx=i
#                             addedIdx=0
#                     else:
#                         prod=nums1[i]*nums2[remIdxs[i][-1]]
#                         if prod<minVal:
#                             minVal=prod
#                             minIdx=i
#                             addedIdx=-1
#             remIdxs[minIdx].pop(addedIdx)
#             lastK=minVal
#             k-=1
#         return lastK
    
# import bisect
# from typing import List
# class Solution:
#     def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
#         def countLessEqual(x):
#             count = 0
#             for a in nums1:
#                 if a > 0: count += bisect.bisect_right(nums2, x // a)
#                 elif a < 0: count += len(nums2) - bisect.bisect_left(nums2, (x + (-1 if x % a else 0)) // a)
#                 elif a == 0 and x >= 0: count += len(nums2)
#             return count
        
#         left = -10**10
#         right = 10**10
#         while left < right:
#             mid = (left + right) // 2
#             if countLessEqual(mid) < k: left = mid + 1
#             else: right = mid
#         return left

import bisect
from typing import List
class Solution:
    def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
        def count_pairs(x: int) -> int:
            count = 0
            for a in nums1:
                if a > 0: count += bisect.bisect_right(nums2, x // a)
                elif a < 0:
                    target = x // a
                    if x % a != 0: target += 1
                    count += len(nums2) - bisect.bisect_left(nums2, target)
                else:
                    if x >= 0: count += len(nums2)
            return count

        low = -10**10
        high = 10**10
        while low < high:
            mid = (low + high) // 2
            if count_pairs(mid) < k: low = mid + 1
            else: high = mid
        return low